blocks/ai-chat/src/complete-chat.tsx

```diff
@@ -0,0 +1,486 @@
+  );
+
+  const createRequestMessageEntity = useMemo(
+    () =>
+      createRequestMessageEntityMethod({ graphModule, aiChatBlockEntityId }),
+    [graphModule, aiChatBlockEntityId],
+  );
+
+  const createResponseMessageEntity = useMemo(
+    () =>
+      createResponseMessageEntityMethod({ graphModule, aiChatBlockEntityId }),
+    [graphModule, aiChatBlockEntityId],
+  );
+
+  const submitUserMessage = useCallback(
+    async (params: {
+      previousResponseId?: ResponseId;
+      userMessage: OpenAIChatMessage<"user">;
+    }) => {
+      const { previousResponseId, userMessage } = params;
+
+      const requestId = createRequestId();
+      const responseId = createResponseId();
+
+      const request: CompleteChatRequest = {
+        id: requestId,
+        message: userMessage,
+        active: true,
+        childResponseIds: [responseId],
+      };
+
+      const previousResponse = completeChatResponses.find(
+        ({ id }) => id === previousResponseId,
+      );
+
+      if (previousResponse && !previousResponse.entityId) {
+        throw new Error("Previous response does not have an entity ID");
+      }
+
+      /**
+       * Intentionally don't `await` the promise so that calling the `completeChat`
+       * method isn't delayed
+       */
+      const promisedRequestMessageEntity = createRequestMessageEntity({
+        parentResponseEntityId: previousResponse?.entityId,
+        messageContent: userMessage.content,
+      });
+
+      const response: CompleteChatResponse = {
+        id: responseId,
+        message: { role: "assistant" },
+        active: true,
+        childRequestIds: [],
+      };
+
+      setCompleteChatRequests((previousRequests) => [
+        ...previousRequests,
+        request,
+      ]);
+
+      if (previousResponseId) {
+        setCompleteChatResponses((previousResponses) => {
+          const previousResponseIndex = previousResponses.findIndex(
+            ({ id }) => id === previousResponseId,
+          );
+
+          if (previousResponseIndex === -1) {
+            throw new Error("Previous response not found");
+          }
+
+          return [
+            ...previousResponses.slice(0, previousResponseIndex),
+            {
+              ...previousResponses[previousResponseIndex]!,
+              childRequestIds: [
+                ...previousResponses[previousResponseIndex]!.childRequestIds,
+                requestId,
+              ],
+            },
+            ...previousResponses.slice(previousResponseIndex + 1),
+          ];
+        });
+      } else {
+        setRootChatRequestId(requestId);
+      }
+
+      setTimeout(() => {
+        setCompleteChatResponses((previousResponses) => {
+          const existingResponse = previousResponses.find(
+            ({ id }) => id === response.id,
+          );
+
+          return existingResponse
+            ? previousResponses
+            : [...previousResponses, response];
+        });
+      }, 500);
```

Comment:
It's so that the response message skeleton isn't immediately displayed alongside the request message, so that their entrance animations are triggered sequentially rather than at the same time.

IMO this creates an effect that's closer to what users expect when having a conversation with another being, as it will not immediately start thinking about how to respond.

Reply:
I've tried to account for the race condition where the completed response may return under 500ms

Reply:
I see, thanks!